src/librustdoc/clean/def_ctor.rs

```diff
@@ -0,0 +1,65 @@
+// Copyright 2018 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+use core::DocContext;
+
+use super::*;
+
+pub fn get_def_ctor_from_def_id<F>(cx: &DocContext,
+                                   def_id: DefId,
+                                   callback: &F,
+) -> Vec<Item>
+where F: Fn(& dyn Fn(DefId) -> Def) -> Vec<Item> {
```

Comment:
The filtering is done before.

Reply:
Also, why would it be better to replace `Vec<Item>` with `R`?

Reply:
Because `Vec<Item>` is *very arbitrary* and the function doesn't have anything to do with `Vec<Item>`, that's just what the closures it's used with deal with.

Returning `Option<Def>` is probably a better idea.

Reply:
This comment is still outstanding, but has been hidden by future commits editing the area around it.

I would prefer to see a generic return value - that's how these callback methods tend to work. However, since there is that arm returning `Vec::new()`, we should either depend on the return value implementing `Default`, or restructure it so that a callback is not necessary.